### PR TITLE
Added option to export in CFF (Citation File Format)

### LINKED ISSUE #10661: Add exporter for Citation File Format (CFF) (yaml based)
**Is your suggestion for improvement related to a problem? Please describe.**
JabRef should be able to export citations into the cff file format. There is already an implemented importer which maps bibtex fields.
The existing YAML exporter can be easily adapted.

**Additional context**

https://citation-file-format.github.io/

An example python converter
https://github.com/citation-file-format/cffconvert

### PR DESCRIPTION
**Pull request by JEANPRBT**
This PR adds the possibility to export BibEntries in CFF format, as requested in #10661. It supports several entries per .cff file, but I don't know if this is permitted by the CFF format. This is hard to solve since the export process in the code is intended for several entries. Also, since the CFF format is primarily intended for pieces of software, this exporter uses credit redirection with the line [`preferred-citation`](https://github.com/citation-file-format/citation-file-format/blob/main/schema-guide.md#credit-redirection) to specify another type of document.
The fields supported are entry type, author's name, abstract, DOI, volume and URL. 

The date is quite complicated since CFF forces date to be of the form "yyyy-mm-dd". This causes problems if only month and year are specified, i.e. 2021-03. It can be solved using a DateFormatter specifically for the CFF exporter and using a default day number, but I did not know if that was necessary, hence I did not implement this. 

Documentation is not up-to-date, CFF as export format must be documented. 

article
new_entry
success
result

Closes #10661 

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by Siedlerchr** [comment]
Thanks for the contribution!

For date, you could checkout the dateformatter, howeer I don't know how it behaves with missing info. Maybe you can add some more test strings for that (DateFormatterTest) 

> DateFormatter : formats a date. With no argument, the date is given in ISO-format (yyyy-MM-dd), which is also the expected format of the input. The argument may contain yyyy, MM, and dd in any combination. For example \format[DateFormatter(MM/yyyy)]{\date} will output 07/2016 if the date field contains 2016-07-15.

https://docs.jabref.org/collaborative-work/export/customexports

**Comment by JEANPRBT** [comment]
I fixed the date and implemented this field when exporting to CFF. 
Now, when a date contains year, month and day, it is displayed as following : 
```
date-released: yyyy-mm-dd
```

It it only contains month and year, it is displayed as follows : 
```
 month: mm
 year: yyyy
```

It is only contains year, it is displayed as follows:  
```
year: yyyy
```

Finally, if it is poorly formatted, is is displayed as follows:
```
issue-date: <date-as-is>
```

This allows not to invent a day and adapt the format to acceptable CFF fields if day or month are missing. Hope you're happy with that !

article
result

**Review by Siedlerchr — DISMISSED** [review]
Looks good to me now, I'm okay with the pragmatic date approach

**Comment by Siedlerchr** [comment]
A bit out of scope, but if you ever need to parse more complicated date formats you can see Jabref's Date class which contains some more crazy formats

File: src/main/java/org/jabref/logic/layout/format/CffDate.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        This class is used to parse dates for CFF exports. Since we do not know if the input String contains
+        year, month and day, we must go through all these cases to return the best CFF format possible.
+        Different cases are stated below.
+
+        Year, Month and Day contained   => date-released: yyyy-mm-dd
+        Year and Month contained        => month: mm
+                                           year: yyyy
+        Year contained                  => year: yyyy
+        Poorly formatted                => issue-date: <fieldText>
+    */
```
Comment (koppor) [inline src/main/java/org/jabref/logic/layout/format/CffDate.java:23]: Please make a class comment -> JAvaDoc. Put it above line 11 and with `/** ... */`.

File: src/main/java/org/jabref/logic/layout/format/CffDate.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        } catch (DateTimeParseException e) {
+            try {
+                formatString = "yyyy-MM";
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern(formatString);
+                YearMonth yearMonth = YearMonth.parse(fieldText, formatter);
+                int month = yearMonth.getMonth().getValue();
+                int year = yearMonth.getYear();
+                builder.append("month: ");
+                builder.append(month);
+                builder.append(System.lineSeparator());
```
Comment (koppor) [inline src/main/java/org/jabref/logic/layout/format/CffDate.java:43]: JabRef has `OS.Newline` (or similar)

File: src/main/java/org/jabref/logic/layout/format/CffDate.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                builder.append("  ");
```
Comment (koppor) [inline src/main/java/org/jabref/logic/layout/format/CffDate.java:44]: Why two spaces before year, but no spaces before `month:`?

If this is really CFF conform, please add a comment. And maybe merge the two lines to

    builder.append("  year: ");

File: src/main/resources/resource/layout/cff.layout [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  authors:
+    - name: \format[Default(No author specified.)]{\author}
+  title: \format[Default(No title specified.)]{\title}
+\begin{date}
+  \format[CffDate]{\date}
+\end{date}
+\begin{abstract}  abstract: \abstract\end{abstract}
+\begin{doi}  doi: \doi\end{doi}
+\begin{volume}  volume: \volume\end{volume}
+\begin{url}  url: "\url"\end{url}
```
Comment (koppor) [inline src/main/resources/resource/layout/cff.layout:17]: Please add a final newline here.

**Review by koppor — CHANGES_REQUESTED** [review]
Thank you for the contribution.

Some small comments.

Do you have time to craft a roundtrip test? We now have an exporter and an importer.

**Comment by JEANPRBT** [comment]
I do not have much time to implement a roundtrip test, but I think this is more complicated than it seems. In fact, the importer only handles two types of entries, `software` and `dataset`. Because the exporter supports a much wider range of entries, using the `preferred-citation` reference block, this is hard to get the same input and output files. 
Furthermore, I did not find any way to extract the first name of an author separately from his last name. This means I am forced to use the field `name` for `authors`, whereas many .cff files use separate `last-name` and `given-names` fields.

**Comment by Siedlerchr** [comment]
I'm okay with that for the moment.  Then we create a new task for roundtrip test.
For author parsing:
https://github.com/JabRef/jabref/blob/79a775665fc0af01306e6eb69e7686d2c5d27ff9/src/main/java/org/jabref/model/entry/AuthorList.java#L165-L172

**Comment by Raahitya-14** [comment]
> Thank you for the contribution.
> 
> Some small comments.
> 
> Do you have time to craft a roundtrip test? We now have an exporter and an importer.

Hey, I am Jean's friend, I can work on crafting a round-trip test! Please let me know if that's fine.

**Comment by koppor** [comment]
@Raahitya-14 Sure, go ahead! Maybe the importer has to be improved, too.

As first test https://github.com/JabRef/jabref/blob/main/CITATION.cff Should be imported. Maybe, your work turns out to improve the importer...

**Review by Siedlerchr — APPROVED** [review]

**Comment by Raahitya-14** [comment]
> @Raahitya-14 I suggest you create a new PR for this then

Sure, I will create a new PR

**Review by koppor — APPROVED** [review]

**Pull request merged**